### ISSUE TITLE
Attempting to pull missing image from insecure registry gives wrong message

### ISSUE THREAD

**Issue by ColinChartier** [containerd/cri#1201]
Before image is pushed:
```
crictl pull 172.17.0.3:31653/webserver-doctest:e7376bfde60d
FATA[0000] pulling image failed: rpc error: code = Unknown desc = failed to resolve image "172.17.0.3:31653/webserver-doctest:e7376bfde60d": no available registry endpoint: failed to do request: Head https://172.17.0.3:31653/v2/webserver-doctest/manifests/e7376bfde60d: http: server gave HTTP response to HTTPS client
```

After image is pushed:
```
crictl pull 172.17.0.3:31653/webserver-doctest:5028ed6cc4ab
Image is up to date for sha256:8866ef28793fc646f5aff3434eb02406831a66ce6efcaac3f34a65642dfb46fb
```

Insecure registry config:
```
        [plugins.cri.registry.mirrors."172.17.0.3:31653"]
          endpoint = ["http://172.17.0.3:31653"]
```

Relevant logs:
```
INFO[2019-07-18T15:08:31.194164353Z] PullImage "172.17.0.3:31653/webserver-doctest:e7376bfde60d" 
DEBU[2019-07-18T15:08:31.194328492Z] resolving                                    
DEBU[2019-07-18T15:08:31.194360171Z] do request                                    request.headers=map[Accept:[application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, *]] request.method=HEAD url="http://172.17.0.3:31653/v2/webserver-doctest/manifests/e7376bfde60d"
DEBU[2019-07-18T15:08:31.196526114Z] fetch response received                       response.headers=map[Content-Type:[application/json; charset=utf-8] Docker-Distribution-Api-Version:[registry/2.0] X-Content-Type-Options:[nosniff] Date:[Thu, 18 Jul 2019 15:08:31 GMT] Content-Length:[102]] status="404 Not Found" url="http://172.17.0.3:31653/v2/webserver-doctest/manifests/e7376bfde60d"
DEBU[2019-07-18T15:08:31.196596998Z] resolving                                    
DEBU[2019-07-18T15:08:31.196612422Z] do request                                    request.headers=map[Accept:[application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, *]] request.method=HEAD url="https://172.17.0.3:31653/v2/webserver-doctest/manifests/e7376bfde60d"
ERRO[2019-07-18T15:08:31.197066387Z] PullImage "172.17.0.3:31653/webserver-doctest:e7376bfde60d" failed  error="failed to resolve image "172.17.0.3:31653/webserver-doctest:e7376bfde60d": no available registry endpoint: failed to do request: Head https://172.17.0.3:31653/v2/webserver-doctest/manifests/e7376bfde60d: http: server gave HTTP response to HTTPS client"
```

It seems that even with an explicitly insecure registry, it still attempts to use https when it receives a 404, and then tries https, and is surprised at getting an http response.

**Comment by mikebrow** [comment]
please provide:
cat /etc/containerd/config.toml
containerd -v

**Comment by ColinChartier** [comment]
```
# cat /etc/containerd/config.toml 
disabled_plugins = ["aufs", "btrfs", "zfs"]
root = "/var/lib/containerd"
state = "/run/containerd"
oom_score = 0

[grpc]
  address = "/run/containerd/containerd.sock"
  uid = 0
  gid = 0
  max_recv_message_size = 16777216
  max_send_message_size = 16777216

[debug]
  address = ""
  uid = 0
  gid = 0
  level = ""

[metrics]
  address = ""
  grpc_histogram = false

[cgroup]
  path = ""

[plugins]
  [plugins.cgroups]
    no_prometheus = false
  [plugins.cri]
    stream_server_address = "127.0.0.1"
    stream_server_port = "0"
    enable_selinux = false
    sandbox_image = "k8s.gcr.io/pause:3.1"
    stats_collect_period = 10
    systemd_cgroup = false
    enable_tls_streaming = false
    max_container_log_line_size = 16384
    [plugins.cri.containerd]
      snapshotter = "overlayfs"
      no_pivot = false
      [plugins.cri.containerd.default_runtime]
        runtime_type = "io.containerd.runtime.v1.linux"
        runtime_engine = ""
        runtime_root = ""
      [plugins.cri.containerd.untrusted_workload_runtime]
        runtime_type = ""
        runtime_engine = ""
        runtime_root = ""
    [plugins.cri.cni]
      bin_dir = "/opt/cni/bin"
      conf_dir = "/etc/cni/net.d"
      conf_template = ""
    [plugins.cri.registry]
      [plugins.cri.registry.mirrors]
        [plugins.cri.registry.mirrors."172.17.0.3:31653"]
          endpoint = ["http://172.17.0.3:31653"]
        [plugins.cri.registry.mirrors."docker.io"]
          endpoint = ["https://registry-1.docker.io"]
    [plugins.cri.x509_key_pair_streaming]
      tls_cert_file = ""
      tls_key_file = ""
  [plugins.diff-service]
    default = ["walking"]
  [plugins.linux]
    shim = "containerd-shim"
    runtime = "runc"
    runtime_root = ""
    no_shim = false
    shim_debug = false
  [plugins.opt]
    path = "/opt/containerd"
  [plugins.restart]
    interval = "10s"
  [plugins.scheduler]
    pause_threshold = 0.02
    deletion_threshold = 0
    mutation_threshold = 100
    schedule_delay = "0s"
    startup_delay = "100ms"
```
# containerd -v
containerd github.com/containerd/containerd 1.2.6-0ubuntu1 
```

**Comment by mikebrow** [comment]
have to think this over... the code checks the mirrors then the default host which it derives from the reference.. And as you can see the default is to use https://reference... 

There is no check in the code for failing from the mirror to the default where the default is derived from whether or not some number of mirrors were http or https and there is no way, to my knowledge, for you to specify what the default should be. 

hmm...

**Comment by Random-Liu** [comment]
Hm, there is an issue here.

We blindly try mirrors one by one no matter what errors we hit, including image not found.
We should only fallback to mirrors when there is a connection problem to the registry. There is a complex logic about this in Docker https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/distribution/errors.go#L116

**Comment by Random-Liu** [comment]
We should use the new registry mirror support in containerd https://github.com/containerd/containerd/pull/3400, which should fix this issue.

**Comment by mikebrow** [comment]
That works.

**Comment by joedborg** [comment]
I'm seeing the same issue with a private registry that's HTTP(notS) and basic auth.

**Comment by joedborg** [comment]
It is working via `ctr`.

```bash
sudo ctr image pull --skip-verify --plain-http --user testuser:testpassword 172.31.55.16:5000/nginx:lates
```

**Comment by Random-Liu** [comment]
Both the original "wrong error message" issue, and the issue mentioned in https://github.com/containerd/cri/issues/1201#issuecomment-514835772 will be fixed in 1.3 by https://github.com/containerd/cri/pull/1227

I changed the milestone to 1.2, because I think we should cherry-pick a small fix for the "wrong error message" in 1.2.

**Comment by joedborg** [comment]
Is this confirmed fixed?  I think I'm still seeing it (1.3.3):

```
[plugins."io.containerd.grpc.v1.cri".registry.mirrors."184.73.131.126:5000"]
  endpoint = ["http://184.73.131.126:5000"]
```

```
Events:
  Type     Reason     Age        From                       Message
  ----     ------     ----       ----                       -------
  Normal   Scheduled  <unknown>  default-scheduler          Successfully assigned default/test to ip-172-31-43-150
  Normal   Pulling    6s         kubelet, ip-172-31-43-150  Pulling image "184.73.131.126:5000/nginx:latest"
  Warning  Failed     6s         kubelet, ip-172-31-43-150  Failed to pull image "184.73.131.126:5000/nginx:latest": rpc error: code = Unknown desc = failed to pull and unpack image "184.73.131.126:5000/nginx:latest": failed to resolve reference "184.73.131.126:5000/nginx:latest": failed to do request: Head https://184.73.131.126:5000/v2/nginx/manifests/latest: http: server gave HTTP response to HTTPS client
  Warning  Failed     6s         kubelet, ip-172-31-43-150  Error: ErrImagePull
  Normal   BackOff    6s         kubelet, ip-172-31-43-150  Back-off pulling image "184.73.131.126:5000/nginx:latest"
  Warning  Failed     6s         kubelet, ip-172-31-43-150  Error: ImagePullBackOff

```

**Comment by nethi** [comment]
> Is this confirmed fixed? I think I'm still seeing it (1.3.3):
> 
> ```
> [plugins."io.containerd.grpc.v1.cri".registry.mirrors."184.73.131.126:5000"]
>   endpoint = ["http://184.73.131.126:5000"]
> ```
> 
> ```
> Events:
>   Type     Reason     Age        From                       Message
>   ----     ------     ----       ----                       -------
>   Normal   Scheduled  <unknown>  default-scheduler          Successfully assigned default/test to ip-172-31-43-150
>   Normal   Pulling    6s         kubelet, ip-172-31-43-150  Pulling image "184.73.131.126:5000/nginx:latest"
>   Warning  Failed     6s         kubelet, ip-172-31-43-150  Failed to pull image "184.73.131.126:5000/nginx:latest": rpc error: code = Unknown desc = failed to pull and unpack image "184.73.131.126:5000/nginx:latest": failed to resolve reference "184.73.131.126:5000/nginx:latest": failed to do request: Head https://184.73.131.126:5000/v2/nginx/manifests/latest: http: server gave HTTP response to HTTPS client
>   Warning  Failed     6s         kubelet, ip-172-31-43-150  Error: ErrImagePull
>   Normal   BackOff    6s         kubelet, ip-172-31-43-150  Back-off pulling image "184.73.131.126:5000/nginx:latest"
>   Warning  Failed     6s         kubelet, ip-172-31-43-150  Error: ImagePullBackOff
> ```

Any update on this ? I am running 1.3.3 (deployed as part of K3d & K3s) and seeing the exact same issue.

**Comment by nethi** [comment]
> > Is this confirmed fixed? I think I'm still seeing it (1.3.3):
> > ```
> > [plugins."io.containerd.grpc.v1.cri".registry.mirrors."184.73.131.126:5000"]
> >   endpoint = ["http://184.73.131.126:5000"]
> > ```
> > 
> > 
> > ```
> > Events:
> >   Type     Reason     Age        From                       Message
> >   ----     ------     ----       ----                       -------
> >   Normal   Scheduled  <unknown>  default-scheduler          Successfully assigned default/test to ip-172-31-43-150
> >   Normal   Pulling    6s         kubelet, ip-172-31-43-150  Pulling image "184.73.131.126:5000/nginx:latest"
> >   Warning  Failed     6s         kubelet, ip-172-31-43-150  Failed to pull image "184.73.131.126:5000/nginx:latest": rpc error: code = Unknown desc = failed to pull and unpack image "184.73.131.126:5000/nginx:latest": failed to resolve reference "184.73.131.126:5000/nginx:latest": failed to do request: Head https://184.73.131.126:5000/v2/nginx/manifests/latest: http: server gave HTTP response to HTTPS client
> >   Warning  Failed     6s         kubelet, ip-172-31-43-150  Error: ErrImagePull
> >   Normal   BackOff    6s         kubelet, ip-172-31-43-150  Back-off pulling image "184.73.131.126:5000/nginx:latest"
> >   Warning  Failed     6s         kubelet, ip-172-31-43-150  Error: ImagePullBackOff
> > ```
> 
> Any update on this ? I am running 1.3.3 (deployed as part of K3d & K3s) and seeing the exact same issue.

For anyone looking for the solution, after some googling, I found that this is now supported with another option in TLS plugin:
https://github.com/containerd/cri/blob/master/docs/registry.md#configure-registry-tls-communication

**Comment by dashjay** [comment]
I gave an answer here https://stackoverflow.com/questions/65724285/kubernetes-with-containerd-http-server-gave-http-response-to-https-client
